### PR TITLE
(WIP) Add a new endpoint that exposes block editor settings through the REST API

### PR DESCRIPTION
**Pull request by Tug**
## Description
This new endpoint at `/__experimental/block-editor-settings` will return all the current settings for the block editor. This will be especially useful for the mobile block editor to set up the GSS (Global Style System) and may even be used to load block patterns.

This is a WIP and will need to be discussed first.

## How has this been tested?
Untested code right now!

## Types of changes
New core REST API endpoint

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: lib/class-wp-rest-block-editor-settings-controller.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	 * Registers the necessary REST API routes.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_settings' ),
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-block-editor-settings-controller.php:33]:
```suggestion
					'callback'            => array( $this, 'get_items' ),
```

File: lib/class-wp-rest-block-editor-settings-controller.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+					'permission_callback' => array( $this, 'get_settings_permissions_check' )
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-block-editor-settings-controller.php:34]:
```suggestion
					'permission_callback' => array( $this, 'get_items_permissions_check' )
```

**Comment by TimothyBJacobs** [comment]
Ideally this would have a schema documenting the response format. Is the block editor settings array structured enough to support that?

**Comment by Tug** [comment]
@TimothyBJacobs Good question! I don't have a precise answer for it right now.

Afaik  "editor settings" on the client is currently defined by its [default value](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/defaults.js#L38).
If we do want to maintain a schema for it, ideally we'd need it at one place so it's common for the backend and the client. I'll see what we can do here 👍

File: lib/class-wp-rest-block-editor-settings-controller.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	 * Return all block editor settings
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_settings( $request ) {
+		$settings = apply_filters( 'block_editor_settings', [] );
```
Comment (spacedmonkey) [inline lib/class-wp-rest-block-editor-settings-controller.php:68]:
```suggestion
		$settings = apply_filters( 'block_editor_settings', array() );
```

**Comment by spacedmonkey** [comment]
This endpoint should be preloaded on the page.

**Review by InfiniteLove2020 — APPROVED** [review]

File: lib/class-wp-rest-block-editor-settings-controller.php [hunk 4]
```diff
@@ -31,22 +31,18 @@ public function register_routes() {
...
 	/**
 	 * Checks whether a given request has permission to read block editor settings
 	 *
-	 * @since 5.5.0
-	 *
-	 * @param WP_REST_Request $request Full details about the request.
-	 *
 	 * @return WP_Error|bool True if the request has permission, WP_Error object otherwise.
 	 */
-	public function get_items_permissions_check( $request ) {
```
Comment (chipsnyder) [inline lib/class-wp-rest-block-editor-settings-controller.php]: @geriux I think You might still need this when I pull in this commit I get:

> Warning: Declaration of WP_REST_Block_Editor_Settings_Controller::get_items() should be compatible with WP_REST_Controller::get_items($request) in /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php on line 59 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-includes/functions.php on line 6362 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-admin/includes/misc.php on line 1310 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-admin/admin-header.php on line 9 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-includes/option.php on line 1051 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-includes/option.php on line 1052

File: lib/class-wp-rest-block-editor-settings-controller.php [hunk 5]
```diff
@@ -58,13 +54,9 @@ public function get_items_permissions_check( $request ) {
 	/**
 	 * Return all block editor settings
 	 *
-	 * @since 5.5.0
-	 *
-	 * @param WP_REST_Request $request Full details about the request.
-	 *
 	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
 	 */
-	public function get_items( $request ) {
```
Comment (chipsnyder) [inline lib/class-wp-rest-block-editor-settings-controller.php]: @geriux I think You might still need this when I pull in this commit I get:

> Warning: Declaration of WP_REST_Block_Editor_Settings_Controller::get_items() should be compatible with WP_REST_Controller::get_items($request) in /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php on line 59 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-includes/functions.php on line 6362 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-admin/includes/misc.php on line 1310 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-admin/admin-header.php on line 9 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-includes/option.php on line 1051 Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-block-editor-settings-controller.php:0) in /var/www/html/wp-includes/option.php on line 1052

**Comment by geriux** [comment]
Closing in favor of https://github.com/WordPress/gutenberg/pull/29969

**Pull request closed without merging**